File: data_structures/trie.py
from collections import defaultdict
from dataclasses import dataclass, field
from typing import DefaultDict, Callable


@dataclass
class TrieNode:
    has_word: bool = False
    children: DefaultDict[str, "TrieNode"] = field(default_factory=defaultdict)


class Trie:
    def __init__(self):
        self._root = TrieNode()

    @property
    def root(self):
        return self._root

    def insert(self, word: str) -> None:
        current = self._root

        for ch in word:
            if ch not in current.children:
                current.children[ch] = TrieNode()
            current = current.children[ch]

        current.has_word = True

    def full_search(self, word: str) -> bool:
        return self._search(word, lambda node: node.has_word)

    def starts_with(self, prefix: str) -> bool:
        return self._search(prefix, lambda _: True)

    def _search(self, word: str, func: Callable[[TrieNode], bool]) -> bool:
        current = self._root

        for ch in word:
            if ch not in current.children:
                return False

            current = current.children[ch]

        return func(current)

File: cp/src/utility/trie.rs
use std::borrow::Borrow;
use std::cell::Cell;
use std::collections::HashMap;

pub struct TrieNode {
    is_word: bool,
    children: Cell<HashMap<char, TrieNode>>,
}

pub struct Trie {
    root: TrieNode,
}

impl Trie {
    fn new() -> Self {
        Trie {
            root: TrieNode {
                is_word: false,
                children: Cell::new(HashMap::new()),
            },
        }
    }

    fn insert(&mut self, word: String) {
        let mut current = &self.root;
        let word_len = word.len();
        for (i, ch) in word.chars().enumerate() {
            let mut map = current.children.borrow().get_mut();
            if Some(node) = map.get(&ch) {
                current = node;
            } else {
                let node = TrieNode {
                    is_word: false,
                    children: Cell::new(HashMap::new()),
                };
                map.insert(ch, node);
                current = &node;
            }
        }

        current.is_word = true;
    }

    fn search(&self, word: String) -> bool {}

    fn starts_with(&self, prefix: String) -> bool {}
}
